### PR TITLE
Fix leading-zero stripping in bn normalize method (makes ECC curves k256 & k192 usable again)

### PR DESCRIPTION
**Pull request by letorbi**
The leading-zero stripping within the normalization method of the bn class breaks the multiplication of points on the ECC curves k192, k224 and k256.

This PR removes the leading-zero stripping from the normalize method, which was introduced in commit ebfab1c. It adds a leading-zero stripping option to the toString method of the bn class instead. This hopefully complies with the original intention of commit ebfab1c to clean up the output of the toString method. Apart from that also a test case for ECC curve point multiplication has been added.

When this PR has been merged, the ECC curves k256 and k192 will be usable again. Curve k224 however seems to have more problems and is still not usable.

### CODE REVIEW

**Comment by letorbi** [comment]
The checks fails due to the still broken k224 curve.

**Comment by Nilos** [comment]
I will not merge PRs which have red tests... Can we try to fix it in this PR too?

File: core/bn.js [hunk 1]
```diff
@@ -88,12 +88,15 @@ sjcl.bn.prototype = {
...
    * Convert to a hex string.
    */
-  toString: function() {
-    this.fullReduce();
-    var out="", i, s, l = this.limbs;
-    for (i=0; i < this.limbs.length; i++) {
+  toString: function(doTrim) {
+    var i, s, l, out = new sjcl.bn(this);
+    out.fullReduce();
+	doTrim && out.trim();
```
Comment (Nilos) [inline core/bn.js]: Whitespace error!

**Comment by Nilos** [comment]
But definite :+1: for the tests and the PR in general!

File: core/bn.js [hunk 2]
```diff
@@ -88,12 +88,15 @@ sjcl.bn.prototype = {
...
+    l = out.limbs;
+	out = "";
```
Comment (Nilos) [inline core/bn.js]: Another whitespace error

Comment (Nilos) [inline core/bn.js]: This also changes the type of the `out`-Variable which makes all major browsers deoptimize the function.

**Comment by letorbi** [comment]
Hehe, yeah. I use tabs for indent and it seems that I've overseen some of them ^^ I'll fix that later this evening. Right know I'm trying to fix curve k224...

About the failing tests: I understand that you won't merge a PR which doesn't pass the tests. However, this is not due to a bug in this PR bug to another bug somewhere else. So I have two options: Either fix the k224 bug (seems to be tricky) or not include k224 in the tests. Would any of these be ok with you?

File: core/bn.js [hunk 3]
```diff
@@ -88,12 +88,15 @@ sjcl.bn.prototype = {
   /**
    * Convert to a hex string.
    */
-  toString: function() {
-    this.fullReduce();
-    var out="", i, s, l = this.limbs;
-    for (i=0; i < this.limbs.length; i++) {
+  toString: function(doTrim) {
+    var i, s, l, out = new sjcl.bn(this);
+    out.fullReduce();
```
Comment (Nilos) [inline core/bn.js]: I am pretty sure that fullReduce also has the same effect as `trim()` or am I missing something hereß

**Comment by Nilos** [comment]
Not including k224 in the tests (the old behaviour so to say) would be fine by me, even though fixing the k224 bug would be the best ;)

File: core/bn.js (hunk 3 above)
Comment (letorbi) [inline core/bn.js]: `fullReduce` seems to be a stub and only calls `this.normalize()`. Actually a think that there's a lot of refactoring and optimization potential in the whole bn code...

**Comment by Nilos** [comment]
Maybe we can restructure the tests and add more test cases in the following way:
Data format:
```
var vectors = [{
  pri: "0x...",
  x: "0x...",
  y: "0x...",
  curve: "c192"
 }]
```
This way we could have many test vectors for all curves and thus find out which values cause problems in a better way.

**Comment by letorbi** [comment]
Well, I went a bit deeper into the rabbit hole, but I was not able to figure out what exactly the problem is with k224. I will create an issue for this tomorrow (have to leave my desk now ^^).

Apart from that I fixed the whitespace errors, refactored the tests as you suggested and replaced my original changes to commit ebfab1c by a version, which sticks more to the original code.

**Comment by Nilos** [comment]
:+1:

**Pull request merged**